# Day 3 – SQL via Python: NYC School Data Exploration
---

## Objective

Using Python and SQL together to explore real-world education data. </br>
Connect to a PostgreSQL database, SQL queries, analyze school-level insights, and present insights in this notebook.

---
## Install Libraries

In [ ]:
#Installs required Python packages for data analysis, database abstraction, and PostgreSQL connectivity within the Jupyter Notebook

%pip install pandas sqlalchemy psycopg2-binary 


[notice] A new release of pip is available: 25.0.1 -> 25.3
[notice] To update, run: pip3 install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


---
## Import Libraries

In [4]:
#Imports pandas for data manipulation and sqlalchemy for database connections and SQL operations

import pandas as pd 
import sqlalchemy as sa 

---
## Database Connection

1. Make a connection 
2. Create an engine

In [5]:
#Defines the database connection URL for accessing a PostgreSQL database with SSL enabled

db_url = 'postgresql://neondb_owner:a9Am7Yy5r9_T7h4OF2GN@ep-falling-glitter-a5m0j5gk-pooler.us-east-2.aws.neon.tech:5432/neondb?sslmode=require'

In [6]:
#Sets up the database infrastructure by initializing the engine and establishing a connection
 
engine = sa.create_engine(db_url) #Creates a SQLAlchemy engine using the specified database connection URL
connection = engine.connect() #Opens a connection to the database via the created engine

In [7]:
#Loads multiple tables from the database into pandas DataFrames for further analysis

directory = pd.read_sql("SELECT * FROM nyc_schools.high_school_directory", engine) #Reads the high school directory table from the database into a DataFrame "directory"
demographics = pd.read_sql("SELECT * FROM nyc_schools.school_demographics", engine) #Reads the school demographics table from the database into a DataFrame "demographics"
safety_report = pd.read_sql("SELECT * FROM nyc_schools.school_safety_report", engine) #Reads the school safety report table from the database into a DataFrame "safety_report"

---
## Initial Data Inspection

### Table: directory

In [ ]:
#Displays the first five rows of the directory DataFrame to quickly inspect its structure and contents

directory.head() 

dbn                                        school_name    borough  \
0  27Q260          Frederick Douglass Academy VI High School     Queens   
1  21K559        Life Academy High School for Film and Music   Brooklyn   
2  16K393     Frederick Douglass Academy IV Secondary School   Brooklyn   
3  08X305                               Pablo Neruda Academy      Bronx   
4  03M485  Fiorello H. LaGuardia High School of Music & A...  Manhattan   

  building_code  phone_number    fax_number  grade_span_min  grade_span_max  \
0          Q465  718-471-2154  718-471-2890             9.0              12   
1          K400  718-333-7750  718-333-7775             9.0              12   
2          K026  718-574-2820  718-574-2821             9.0              12   
3          X450  718-824-1682  718-824-1663             9.0              12   
4          M485  212-496-0700  212-724-5748             9.0              12   

   expgrade_span_min  expgrade_span_max start_time end_time  \
0                NaN                NaN    7:45 AM  2:05 PM   
1                NaN                NaN    8:15 AM  3:00 PM   
2                NaN                NaN    8:00 AM  2:20 PM   
3                NaN                NaN    8:00 AM  3:50 PM   
4                NaN                NaN    8:00 AM  4:00 PM   

                                          priority01  \
0  Priority to Queens students or residents who a...   
1  Priority to New York City residents who attend...   
2                 Priority to continuing 8th graders   
3  Priority to Bronx students or residents who at...   
4                    Open to New York City residents   

                                          priority02  \
0  Then to New York City residents who attend an ...   
1                    Then to New York City residents   
2  Then to Brooklyn students or residents who att...   
3  Then to New York City residents who attend an ...   
4  Admission is based on the outcome of a competi...   

                                          priority03  \
0               Then to Queens students or residents   
1                                                      
2  Then to New York City residents who attend an ...   
3                Then to Bronx students or residents   
4  Students must audition for each program (studi...   

                                          priority04  \
0                    Then to New York City residents   
1                                                      
2             Then to Brooklyn students or residents   
3                    Then to New York City residents   
4  Students must be residents of New York City at...   

                        priority05 priority06 priority07 priority08  \
0                                                                     
1                                                                     
2  Then to New York City residents                                    
3                                                                     
4                                                                     

  priority09 priority10 location phone_number2 school_email  \
0                           None          None         None   
1                           None          None         None   
2                           None          None         None   
3                           None          None         None   
4                           None          None         None   

                                        website  \
0  http://schools.nyc.gov/schoolportals/27/Q260   
1  http://schools.nyc.gov/schoolportals/21/K559   
2  http://schools.nyc.gov/schoolportals/16/K393   
3                    www.pablonerudaacademy.org   
4                           www.laguardiahs.org   

                                              subway  \
0                       A to Beach 25th St-Wavecrest   
1  D to 25th Ave ; N to Ave U ; N to Gravesend - ...   
2             J to Kosciusko St ; M, Z to Myrtle Ave   
3               

In [ ]:
#Displays a concise summary of the directory DataFrame, including column names, data types, and non-null counts

directory.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 435 entries, 0 to 434
Columns: 105 entries, dbn to Police Precincts
dtypes: float64(6), int64(1), object(98)
memory usage: 357.0+ KB


In [ ]:
#Generates descriptive statistics for the numeric columns in the directory DataFrame

directory.describe() 

grade_span_min  grade_span_max  expgrade_span_min  expgrade_span_max  \
count      432.000000      435.000000          31.000000          33.000000   
mean         8.442130       11.894253           8.516129          12.363636   
std          1.164591        0.421583           1.121635           0.783349   
min          6.000000        9.000000           6.000000          12.000000   
25%          9.000000       12.000000           9.000000          12.000000   
50%          9.000000       12.000000           9.000000          12.000000   
75%          9.000000       12.000000           9.000000          12.000000   
max          9.000000       12.000000           9.000000          14.000000   

       total_students           bin           bbl  
count      426.000000  4.310000e+02  4.310000e+02  
mean       703.842723  2.581724e+06  2.510921e+09  
std        775.870436  1.189750e+06  1.136455e+09  
min         50.000000  1.000811e+06  1.000160e+09  
25%        349.000000  2.000992e+06  2.023060e+09  
50%        460.500000  2.116159e+06  2.053680e+09  
75%        622.000000  3.330710e+06  3.068830e+09  
max       5458.000000  5.149609e+06  5.066130e+09

In [ ]:
#Calculates the number of missing (null) values in each column of the directory DataFrame

directory.isnull().sum() 

dbn                       0
school_name               0
borough                   0
building_code             0
phone_number              0
                         ..
Zip Codes                 0
Community Districts       0
Borough Boundaries        0
City Council Districts    0
Police Precincts          0
Length: 105, dtype: int64

In [ ]:
#Counts the number of duplicate rows in the directory DataFrame

directory.duplicated().sum() 

np.int64(0)

---
### Table: demographics

In [7]:
#Inspects the demographics DataFrame and adjusts display settings to show all columns

demographics.head() #Displays the first few rows of the demographics DataFrame to preview its structure
pd.set_option('display.max_columns', None) #Configures pandas to display all columns and then outputs the full demographics DataFrame
demographics #Show Output

dbn                                              Name  schoolyear  \
0    01M015                         P.S. 015 ROBERTO CLEMENTE    20052006   
1    01M015                         P.S. 015 ROBERTO CLEMENTE    20062007   
2    01M015                         P.S. 015 ROBERTO CLEMENTE    20072008   
3    01M015                         P.S. 015 ROBERTO CLEMENTE    20082009   
4    01M015                         P.S. 015 ROBERTO CLEMENTE    20092010   
..      ...                                               ...         ...   
238  01M064                               PS 064 ROBERT SIMON    20072008   
239  01M064                               PS 064 ROBERT SIMON    20082009   
240  01M064                               PS 064 ROBERT SIMON    20092010   
241  01M064                               PS 064 ROBERT SIMON    20102011   
242  01M064  PS 064 ROBERT SIMON                                 20112012   

    fl_percent  frl_percent  total_enrollment prek   k grade1 grade2 grade3  \
0         89.4          NaN               281   15  36     40     33     38   
1         89.4          NaN               243   15  29     39     38     34   
2         89.4          NaN               261   18  43     39     36     38   
3         89.4          NaN               252   17  37     44     32     34   
4                      96.5               208   16  40     28     32     30   
..         ...          ...               ...  ...  ..    ...    ...    ...   
238       79.3          NaN               295   26  48     49     47     36   
239       79.3          NaN               300   26  44     48     53     47   
240                    90.5               292   26  46     43     44     51   
241                    90.5               322   32  53     47     45     40   
242                    76.9               324   27  54     50     42     49   

    grade4 grade5 grade6 grade7 grade8 grade9 grade10 grade11 grade12  \
0       52     29     38                                                
1       42     46                                                       
2       47     40                                                       
3       39     49                                                       
4       24     38                                                       
..     ...    ...    ...    ...    ...    ...     ...     ...     ...   
238     46     43                                                       
239     38     44                                                       
240     50     32                                                       
241     52     53                                                       
242     48     54                                                       

     ell_num  ell_percent  sped_num  sped_percent  ctt_num  selfcontained_num  \
0       36.0         12.8        57          20.3     25.0                9.0   
1       38.0         15.6        55          22.6     19.0               15.0   
2       52.0         19.9        60          23.0     20.0               14.0   
3       48.0         19.0        62          24.6     21.0               17.0   
4       40.0         19.2        46          22.1     14.0               14.0   
..       ...          ...       ...           ...      ...                ...   
238     17.0          5.8        68          23.1     37.0               11.0   
239     15.0          5.0        73          24.3     43.0               22.0   
240     13.0          4.5        88          30.1     49.0               24.0   
241     21.0          6.5        98          30.4     64.0               18.0   
242     29.0          9.0       102          31.5     55.0               28.0   

     asian_num  asian_per  black_num  black_per  hispanic_num  hispanic_per  \
0           10        3.6         74       26.3           189          67.3   
1           18        7.4         68       28.0           153          63.0   
2           16        6.1         77       29.5

In [ ]:
#Displays a concise summary of the demographics DataFrame, including column names, data types, and non-null counts

demographics.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 243 entries, 0 to 242
Data columns (total 38 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   dbn                243 non-null    object 
 1   Name               243 non-null    object 
 2   schoolyear         243 non-null    int64  
 3   fl_percent         243 non-null    object 
 4   frl_percent        104 non-null    float64
 5   total_enrollment   243 non-null    int64  
 6   prek               243 non-null    object 
 7   k                  243 non-null    object 
 8   grade1             243 non-null    object 
 9   grade2             243 non-null    object 
 10  grade3             243 non-null    object 
 11  grade4             243 non-null    object 
 12  grade5             243 non-null    object 
 13  grade6             243 non-null    object 
 14  grade7             243 non-null    object 
 15  grade8             243 non-null    object 
 16  grade9             243 non

In [ ]:
#Generates descriptive statistics for the numeric columns in the demographics DataFrame

demographics.describe() 

schoolyear  frl_percent  total_enrollment     ell_num  ell_percent  \
count  2.430000e+02   104.000000        243.000000  239.000000   243.000000   
mean   2.008197e+07    77.683654        389.024691   52.167364    11.863374   
std    2.015634e+04    18.743165        213.035653   84.805296    13.838320   
min    2.005201e+07    19.600000         78.000000    1.000000     0.000000   
25%    2.006201e+07    71.725000        246.500000   13.500000     5.000000   
50%    2.008201e+07    82.050000        338.000000   30.000000     8.500000   
75%    2.010201e+07    91.200000        476.500000   49.500000    13.100000   
max    2.011201e+07    99.700000       1613.000000  486.000000    87.500000   

         sped_num  sped_percent     ctt_num  selfcontained_num   asian_num  \
count  243.000000    243.000000  200.000000         211.000000  243.000000   
mean    67.823045     19.432510   28.275000          19.033175   76.477366   
std     33.497958      7.909256   23.985954          11.881712  121.213537   
min      1.000000      0.200000    0.000000           0.000000    1.000000   
25%     47.000000     15.950000    9.750000          10.000000   17.000000   
50%     67.000000     21.700000   21.000000          17.000000   32.000000   
75%     92.500000     24.050000   43.000000          27.000000   60.000000   
max    166.000000     35.500000  114.000000          58.000000  559.000000   

       ...   black_num   black_per  hispanic_num  hispanic_per   white_num  \
count  ...  243.000000  243.000000    243.000000    243.000000  243.000000   
mean   ...   70.991770   20.324280    189.390947     52.103292   44.362140   
std    ...   34.196925    8.140595    103.313130     18.892945  105.479506   
min    ...   12.000000    2.100000     15.000000      2.900000    0.000000   
25%    ...   44.500000   14.900000    116.000000     45.450000    7.000000   
50%    ...   68.000000   20.600000    171.000000     57.700000   12.000000   
75%    ...   87.000000   25.950000    269.500000     65.100000   31.500000   
max    ...  192.000000   45.400000    503.000000     80.100000  725.000000   

        white_per    male_num    male_per  female_num  female_per  
count  243.000000  243.000000  243.000000  243.000000  243.000000  
mean     9.521399  197.872428   51.502881  191.152263   48.497119  
std     13.164749  103.299753    5.181933  113.588160    5.181933  
min      0.000000   37.000000   25.100000   41.000000   36.200000  
25%      2.000000  128.500000   48.750000  121.000000   45.400000  
50%      3.400000  175.000000   51.700000  173.000000   48.300000  
75%      8.500000  242.500000   54.600000  229.500000   51.250000  
max     56.800000  794.000000   63.800000  819.000000   74.900000  

[8 rows x 21 columns]

# Table Overview

In [ ]:
#Returns a list of all column names in the directory DataFrame

list(directory.columns)

['dbn',
 'school_name',
 'borough',
 'building_code',
 'phone_number',
 'fax_number',
 'grade_span_min',
 'grade_span_max',
 'expgrade_span_min',
 'expgrade_span_max',
 'start_time',
 'end_time',
 'priority01',
 'priority02',
 'priority03',
 'priority04',
 'priority05',
 'priority06',
 'priority07',
 'priority08',
 'priority09',
 'priority10',
 'location',
 'phone_number2',
 'school_email',
 'website',
 'subway',
 'bus',
 'grades2018',
 'finalgrades',
 'total_students',
 'extracurricular_activities',
 'school_sports',
 'attendance_rate',
 'pct_stu_enough_variety',
 'pct_stu_safe',
 'school_accessibility_description',
 'directions1',
 'requirement1',
 'requirement2',
 'requirement3',
 'requirement4',
 'requirement5',
 'program1',
 'code1',
 'interest1',
 'method1',
 'seats9ge1',
 'grade9gefilledflag1',
 'grade9geapplicants1',
 'seats9swd1',
 'grade9swdfilledflag1',
 'grade9swdapplicants1',
 'campus_name',
 'building_borough',
 'building_location',
 'latitude',
 'longitude',
 'community_

In [ ]:
#Returns a list of all column names in the demographics DataFrame

list(demographics.columns)

['dbn',
 'Name',
 'schoolyear',
 'fl_percent',
 'frl_percent',
 'total_enrollment',
 'prek',
 'k',
 'grade1',
 'grade2',
 'grade3',
 'grade4',
 'grade5',
 'grade6',
 'grade7',
 'grade8',
 'grade9',
 'grade10',
 'grade11',
 'grade12',
 'ell_num',
 'ell_percent',
 'sped_num',
 'sped_percent',
 'ctt_num',
 'selfcontained_num',
 'asian_num',
 'asian_per',
 'black_num',
 'black_per',
 'hispanic_num',
 'hispanic_per',
 'white_num',
 'white_per',
 'male_num',
 'male_per',
 'female_num',
 'female_per']

In [ ]:
#Returns a list of all column names in the safety_report DataFrame

list(safety_report.columns)

['school_year',
 'building_code',
 'dbn',
 'location_name',
 'location_code',
 'address',
 'borough',
 'geographical_district_code',
 'register',
 'building_name',
 'num_schools',
 'schools_in_building',
 'major_n',
 'oth_n',
 'nocrim_n',
 'prop_n',
 'vio_n',
 'engroupa',
 'rangea',
 'avgofmajor_n',
 'avgofoth_n',
 'avgofnocrim_n',
 'avgofprop_n',
 'avgofvio_n',
 'borough_name',
 'postcode',
 'latitude',
 'longitude',
 'community_board',
 'council_district',
 'census_tract',
 'bin',
 'bbl',
 'nta',
 '_schools']

# Questions

#### School Distribution

##### How many schools are there in each borough?

In [13]:
#Defines and executes an SQL query to aggregate the number of schools per borough based on joined tables
#Creates a multi-line SQL query that counts distinct schools per borough and orders the results by count

query = """
        SELECT 
          hsd.borough AS borough,
          COUNT(DISTINCT(hsd.dbn)) AS cnt_schools
        FROM nyc_schools.high_school_directory AS hsd
        JOIN nyc_schools.school_safety_report AS ssr
        ON hsd.dbn = ssr.dbn
        GROUP BY hsd.borough
        ORDER BY cnt_schools DESC;
        """

pd.read_sql(query, engine) #Executes the SQL query against the database engine and returns the result as a pandas DataFrame

borough  cnt_schools
0       Brooklyn          121
1          Bronx          118
2      Manhattan          106
3         Queens           80
4  Staten Island           10

#### Language Learners

##### What is the average % of English Language Learners (ELL) per borough?

In [ ]:
#Defines and executes an SQL query to calculate the average percentage of English Language Learners per borough
#Creates a multi-line SQL query that joins school directory and demographics data and orders results by average percentage

query = """
        SELECT
          hsd.borough,
          AVG(sd.ell_percent) AS avg_perc_ell
        FROM nyc_schools.high_school_directory hsd
        LEFT JOIN nyc_schools.school_demographics sd
          ON TRIM(UPPER(hsd.dbn)) = TRIM(UPPER(sd.dbn))
        GROUP BY hsd.borough
        ORDER BY avg_perc_ell ASC;
        """ 

pd.read_sql(query, engine) #Executes the SQL query against the database engine and returns the result as a pandas DataFrame

borough  avg_perc_ell
0      Manhattan        7.5725
1         Queens           NaN
2       Brooklyn           NaN
3  Staten Island           NaN
4          Bronx           NaN

#### School supporting special needs

##### Write a query to find the top 3 schools in each borough with the highest percentage of special education students

In [10]:
#Defines and executes an SQL query to identify the top three schools per borough by special education percentage
#Creates a multi-step SQL query using CTEs to aggregate, rank, and filter schools within each borough

query = """
        WITH per_school AS (
          SELECT
            TRIM(UPPER(dbn)) AS dbn,
            sped_percent
          FROM nyc_schools.school_demographics
          WHERE sped_percent IS NOT NULL
        ),
        ranked AS (
          SELECT
            hsd.borough,
            ps.sped_percent,
            ROW_NUMBER() OVER (PARTITION BY hsd.borough ORDER BY ps.sped_percent DESC) AS rank
          FROM nyc_schools.high_school_directory AS hsd
          JOIN per_school AS ps
            ON TRIM(UPPER(hsd.dbn)) = ps.dbn
          WHERE hsd.borough IS NOT NULL
        )
        SELECT
          borough,
          sped_percent,
          rank
        FROM ranked
        WHERE rank <= 3
        ORDER BY borough, rank;
        """

pd.read_sql(query, engine) #Executes the SQL query against the database engine and returns the result as a pandas DataFrame

borough  sped_percent  rank
0  Manhattan          28.8     1
1  Manhattan          27.7     2
2  Manhattan          26.7     3

# Summary

##### Question 1 

How many schools are there in each borough?

The results show that the number of high schools varies substantially across boroughs, ranging from **10 to 121 schools**. **Brooklyn (121), the Bronx (118), and Manhattan (106)** have the highest counts, while **Queens (80)** and especially **Staten Island (10)** have significantly fewer schools.

##### Question 2 

What is the average % of English Language Learners (ELL) per borough?

The results show that **Manhattan** has an average **ELL percentage of 7.57%**, while **no valid ELL averages are available** for **Queens, Brooklyn, Staten Island, and the Bronx** due to missing or non-aggregable data in the demographics table.

##### Question 3 

Find the top 3 schools in each borough with the highest percentage of special education students

The results show that only **Manhattan** has sufficient data to identify the top three schools by special education percentage, with values ranging from **25.1% to 28.8%**. No ranked results are returned for the other boroughs due to missing or unavailable special education data.